File: app/src/main/java/com/example/repository/DataRepository.java
package com.example.repository;

import android.content.Context;
import androidx.room.Room;
import com.example.model.QAItem;
import com.example.model.QuestItem;
import com.example.sqlite.room.*;
import org.json.JSONArray;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class DataRepository {

    private AppDatabase db;
    private AppDao dao;
    private Context context;
    private static final String DAILY_STATE_PREFS = "StudyMentorDailyState";
    private static final String MAP_PROGRESS_PREFS = "StudyMentorMapProgress";

    public DataRepository(Context context) {
        this.context = context.getApplicationContext();
        db = Room.databaseBuilder(context.getApplicationContext(), AppDatabase.class, "AIStudyMentorRoom.db")
                .fallbackToDestructiveMigration()
                .allowMainThreadQueries()
                .build();
        dao = db.appDao();
        seedInitialData();
    }

    private void seedInitialData() {
        new Thread(() -> {
            List<String> existing = dao.getActiveChapters();
            if (existing == null || existing.isEmpty()) {
                try {
                    java.io.InputStream is = context.getAssets().open("questions.json");
                    int size = is.available();
                    byte[] buffer = new byte[size];
                    is.read(buffer);
                    is.close();
                    String json = new String(buffer, "UTF-8");
                    org.json.JSONObject root = new org.json.JSONObject(json);

                    String[] levels = {"Middle School", "High School", "University / College"};
                    
                    for (String level : levels) {
                        for (java.util.Iterator<String> it = root.keys(); it.hasNext(); ) {
                            String subject = it.next();
                            org.json.JSONObject chaptersObj = root.getJSONObject(subject);

                            int cIndex = 1;
                            for (java.util.Iterator<String> itC = chaptersObj.keys(); itC.hasNext(); ) {
                                String chapterTitle = itC.next();
                                org.json.JSONArray questionsArr = chaptersObj.getJSONArray(chapterTitle);

                                ChapterEntity chapter = new ChapterEntity();
                                chapter.level = level;
                                chapter.subject = subject;
                                chapter.title = chapterTitle;
                                chapter.orderNo = cIndex++;
                                chapter.isActive = 1;
                                long chapterId = dao.insertChapter(chapter);

                                for (int i = 0; i < questionsArr.length(); i++) {
                                    org.json.JSONObject qObj = questionsArr.getJSONObject(i);
                                    String qText = qObj.getString("question");
                                    String optionsJson = qObj.getJSONArray("options").toString();
                                    int correctIdx = Integer.parseInt(qObj.getString("correctAnswer"));
                                    insertQuestion(chapterId, qText, optionsJson, correctIdx);
                                }
                            }
                        }
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }

            List<UserProfileEntity> lb = dao.getLeaderboard();
            if (lb == null || lb.size() < 10) {
                String[] names = {"Alex", "Jordan", "Taylor", "Morgan", "Casey", "Riley", "Jamie", "Skyler", "Cameron", "Quinn"};
                int[] xps = {15200, 14500, 13100, 12500, 11000, 9500, 8000, 7500, 6000, 5000};
                for (int i = 0; i < 10; i++) {
                    UserEntity dummy = new UserEntity();
                    dummy.email = "dummy" + i + "@test.com";
                    dummy.passwordHash = "xxx";
                    long id = dao.insertUser(dummy);
                    if (id != -1) {
                        UserProfileEntity p = new UserProfileEntity();
                        p.userId = id;
                        p.displayName = names[i];
                        p.totalXp = xps[i];
                        dao.insertUserProfile(p);
                    }
                }
            }
        }).start();
    }



    private void insertQuestion(long chapterId, String text, String optionsJson, int correctIdx) {
        QuestionEntity q = new QuestionEntity();
        q.chapterId = chapterId;
        q.questionText = text;
        q.optionsJson = optionsJson;
        q.correctAnswer = String.valueOf(correctIdx);
        dao.insertQuestion(q);
    }

    public List<String> getChaptersByPreference(String level, String subject) {
        return dao.getChaptersByPreference(level, subject);
    }

    public long getChapterIdByTitle(String title) {
        Long id = dao.getChapterIdByTitle(title);
        return id != null ? id : -1;
    }

    public List<QuestItem> getQuestionsByChapter(long chapterId) {
        List<QuestionEntity> questions = dao.getQuestionsByChapter(chapterId);
        List<QuestItem> items = new ArrayList<>();
        for (QuestionEntity q : questions) {
            try {
                JSONArray array = new JSONArray(q.optionsJson);
                List<String> options = new ArrayList<>();
                for (int i = 0; i < array.length(); i++) {
                    options.add(array.getString(i));
                }
                items.add(new QuestItem(q.questionText, options, Integer.parseInt(q.correctAnswer)));
            } catch (Exception e) { e.printStackTrace(); }
        }
        return items;
    }

    public long registerUser(String name, String email, String passwordHash) {
        String normalizedEmail = email == null ? "" : email.trim().toLowerCase(Locale.ROOT);
        Long existingId = dao.getUserIdByEmail(normalizedEmail);
        if (existingId != null) return -2;

        UserEntity user = new UserEntity();
        user.email = normalizedEmail;
        user.passwordHash = passwordHash;
        long userId = dao.insertUser(user);

        if (userId != -1) {
            UserProfileEntity profile = new UserProfileEntity();
            profile.userId = userId;
            profile.displayName = name;
            dao.insertUserProfile(profile);
        }
        return userId;
    }

    public UserEntity login(String email, String passwordHash) {
        String normalizedEmail = email == null ? "" : email.trim().toLowerCase(Locale.ROOT);
        return dao.loginUser(normalizedEmail, passwordHash);
    }

    public long getUserIdByEmail(String email) {
        String normalizedEmail = email == null ? "" : email.trim().toLowerCase(Locale.ROOT);
        Long id = dao.getUserIdByEmail(normalizedEmail);
        return id != null ? id : -1;
    }

    public UserProfileEntity getUserProfile(long userId) {
        return dao.getUserProfile(userId);
    }

    public List<String> getActiveChapters() {
        return dao.getActiveChapters();
    }

    public void saveAIQuestion(long userId, String question, String imageUri, String aiAnswer) {
        AIQuestionEntity ai = new AIQuestionEntity();
        ai.userId = userId;
        ai.questionText = question;
        ai.imageUri = imageUri;
        ai.aiAnswer = aiAnswer;
        ai.createdAt = new java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss", java.util.Locale.US).format(new java.util.Date());
        dao.insertAIQuestion(ai);
    }

    public List<QAItem> getAIHistory(long userId) {
        List<AIQuestionEntity> list = dao.getAIHistory(userId);
        List<QAItem> items = new ArrayList<>();
        for (AIQuestionEntity ai : list) {
            items.add(new QAItem((int)ai.aiQuestionId, ai.questionText, ai.imageUri, ai.aiAnswer, 0));
        }
        return items;
    }

    public void updateXP(long userId, int xpGain) {
        dao.updateXP(userId, xpGain);
    }

    public void updateMana(long userId, int manaGain) {
        dao.updateMana(userId, manaGain);
    }

    public List<DailyTaskEntity> getOrCreateDailyTasks(long userId, String taskDate) {
        List<DailyTaskEntity> tasks = dao.getDailyTasks(userId, taskDate);
        if (tasks != null && !tasks.isEmpty()) return tasks;

        insertDailyTask(userId, taskDate, "LESSON_COMPLETE", 1, 50);
        insertDailyTask(userId, taskDate, "PERFECT_SCORE", 1, 100);
        insertDailyTask(userId, taskDate, "ANSWER_QUESTIONS", 10, 50);
        return dao.getDailyTasks(userId, taskDate);
    }

    private void insertDailyTask(long userId, String taskDate, String taskType, int targetValue, int rewardXp) {
        DailyTaskEntity task = new DailyTaskEntity();
        task.userId = userId;
        task.taskDate = taskDate;
        task.taskType = taskType;
        task.title = taskType;
        task.targetValue = targetValue;
        task.currentValue = 0;
        task.rewardXp = rewardXp;
        task.status = "IN_PROGRESS";
        dao.insertDailyTask(task);
    }

    public void updateDailyTaskProgress(long userId, String taskDate, String taskType, int increment) {
        List<DailyTaskEntity> tasks = dao.getDailyTasks(userId, taskDate);
        if (tasks == null) return;
        for (DailyTaskEntity task : tasks) {
            if (taskType.equals(task.taskType)) {
                dao.updateTaskProgress(task.taskId, increment);
                return;
            }
        }
    }

    public DailyTaskEntity getDailyTask(long taskId) {
        return dao.getDailyTask(taskId);
    }

    public boolean claimDailyTask(long taskId) {
        return dao.claimDailyTask(taskId) > 0;
    }

    private String chestKey(long userId, String taskDate) {
        return "chest_" + userId + "_" + taskDate;
    }

    public boolean isDailyChestClaimed(long userId, String taskDate) {
        return context.getSharedPreferences(DAILY_STATE_PREFS, Context.MODE_PRIVATE)
                .getBoolean(chestKey(userId, taskDate), false);
    }

    public synchronized boolean claimDailyChest(long userId, String taskDate) {
        android.content.SharedPreferences preferences = context.getSharedPreferences(
                DAILY_STATE_PREFS, Context.MODE_PRIVATE);
        String key = chestKey(userId, taskDate);
        if (preferences.getBoolean(key, false)) return false;
        return preferences.edit().putBoolean(key, true).commit();
    }

    private String mapNodeKey(long userId, String chapterTitle) {
        return "node_" + userId + "_" + chapterTitle;
    }

    public int getCurrentLessonNode(long userId, String chapterTitle) {
        return context.getSharedPreferences(MAP_PROGRESS_PREFS, Context.MODE_PRIVATE)
                .getInt(mapNodeKey(userId, chapterTitle), 1);
    }

    public void saveCurrentLessonNode(long userId, String chapterTitle, int node) {
        int safeNode = Math.max(1, Math.min(6, node));
        context.getSharedPreferences(MAP_PROGRESS_PREFS, Context.MODE_PRIVATE)
                .edit()
                .putInt(mapNodeKey(userId, chapterTitle), safeNode)
                .apply();
    }

    public List<UserProfileEntity> getLeaderboard() {
        return dao.getLeaderboard();
    }
}

File: app/src/main/java/com/example/repository/DataRepository.kt
package com.example.repository

import android.content.Context
import com.example.model.QAItem
import com.example.model.QuestItem
import com.example.security.PasswordHasher
import com.example.sqlite.room.*
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class DataRepository @Inject constructor(
    private val dao: AppDao,
    @param:ApplicationContext private val context: Context
) {
    private val scope = CoroutineScope(Dispatchers.IO)

    private val seedJob = scope.launch {
        seedInitialData()
    }

    suspend fun ensureSeeded() {
        seedJob.join()
    }

    private suspend fun seedInitialData() = withContext(Dispatchers.IO) {
        try {
            val existing = dao.getActiveChapters()
            val hasEnglish = existing.any { it.contains("English", ignoreCase = true) || it.contains("Vocabulary", ignoreCase = true) }
            if (!hasEnglish || existing.isEmpty()) {
                dao.clearQuestions()
                dao.clearChapters()

                val jsonString = context.assets.open("questions.json").bufferedReader().use { it.readText() }
                val root = JSONObject(jsonString)
                val levels = arrayOf("Beginner", "Intermediate", "Advanced", "Middle School", "High School", "University / College")

                for (level in levels) {
                    val subjectKeys = root.keys()
                    while (subjectKeys.hasNext()) {
                        val subject = subjectKeys.next()
                        val chaptersObj = root.getJSONObject(subject)
                        var cIndex = 1

                        val chapterKeys = chaptersObj.keys()
                        while (chapterKeys.hasNext()) {
                            val chapterTitle = chapterKeys.next()
                            val questionsArr = chaptersObj.getJSONArray(chapterTitle)

                            val chapter = ChapterEntity(
                                title = chapterTitle,
                                level = level,
                                subject = subject,
                                orderNo = cIndex++,
                                isActive = 1
                            )
                            val chapterId = dao.insertChapter(chapter)

                            for (i in 0 until questionsArr.length()) {
                                val qObj = questionsArr.getJSONObject(i)
                                val qText = qObj.getString("question")
                                val optionsJson = qObj.getJSONArray("options").toString()
                                val correctIdx = qObj.getString("correctAnswer")

                                val question = QuestionEntity(
                                    chapterId = chapterId,
                                    questionText = qText,
                                    optionsJson = optionsJson,
                                    correctAnswer = correctIdx
                                )
                                dao.insertQuestion(question)
                            }
                        }
                    }
                }
            }

            val lb = dao.getLeaderboard()
            if (lb.size < 10) {
                val names = arrayOf("Alex", "Jordan", "Taylor", "Morgan", "Casey", "Riley", "Jamie", "Skyler", "Cameron", "Quinn")
                val xps = intArrayOf(15200, 14500, 13100, 12500, 11000, 9500, 8000, 7500, 6000, 5000)
                for (i in names.indices) {
                    val dummy = UserEntity(
                        email = "dummy$i@test.com",
                        passwordHash = PasswordHasher.hash("kikiDummyPass123!")
                    )
                    val id = dao.insertUser(dummy)
                    if (id != -1L) {
                        val profile = UserProfileEntity(
                            userId = id,
                            displayName = names[i],
                            totalXp = xps[i]
                        )
                        dao.insertUserProfile(profile)
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    suspend fun getChaptersByPreference(level: String?, subject: String?): List<String> = withContext(Dispatchers.IO) {
        ensureSeeded()
        val targetSubject = if (subject.isNullOrBlank() || subject != "English") "English" else subject
        val targetLevel = level ?: "Beginner"
        val results = dao.getChaptersByPreference(targetLevel, targetSubject)
        if (results.isNotEmpty()) {
            results.distinct()
        } else {
            dao.getActiveChapters().distinct()
        }
    }

    suspend fun getChapterIdByTitle(title: String): Long = withContext(Dispatchers.IO) {
        ensureSeeded()
        dao.getChapterIdByTitle(title) ?: -1L
    }

    suspend fun getQuestionsByChapter(chapterId: Long): List<QuestItem> = withContext(Dispatchers.IO) {
        val questions = dao.getQuestionsByChapter(chapterId)
        questions.mapNotNull { q ->
            try {
                val array = JSONArray(q.optionsJson)
                val options = mutableListOf<String>()
                for (i in 0 until array.length()) {
                    options.add(array.getString(i))
                }
                QuestItem(
                    question = q.questionText,
                    options = options,
                    correctIndex = q.correctAnswer.toIntOrNull() ?: 0
                )
            } catch (e: Exception) {
                null
            }
        }
    }

    suspend fun registerUser(name: String, email: String, rawPassword: String): Long = withContext(Dispatchers.IO) {
        val normalizedEmail = email.trim().lowercase(Locale.ROOT)
        val existingId = dao.getUserIdByEmail(normalizedEmail)
        if (existingId != null) return@withContext -2L

        val hashedPassword = PasswordHasher.hash(rawPassword)
        val user = UserEntity(
            email = normalizedEmail,
            passwordHash = hashedPassword,
            createdAt = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US).format(Date())
        )
        val userId = dao.insertUser(user)

        if (userId != -1L) {
            val profile = UserProfileEntity(
                userId = userId,
                displayName = name
            )
            dao.insertUserProfile(profile)
        }
        userId
    }

    suspend fun login(email: String, rawPassword: String): UserEntity? = withContext(Dispatchers.IO) {
        val normalizedEmail = email.trim().lowercase(Locale.ROOT)
        val user = dao.getUserByEmail(normalizedEmail) ?: return@withContext null

        if (PasswordHasher.verify(rawPassword, user.passwordHash)) {
            return@withContext user
        }

        // Tương thích ngược: nếu tài khoản cũ chưa băm mật khẩu
        if (user.passwordHash == rawPassword) {
            // Tự động nâng cấp sang mã băm bảo mật
            user.passwordHash = PasswordHasher.hash(rawPassword)
            dao.insertUser(user)
            return@withContext user
        }

        null
    }

    suspend fun getUserIdByEmail(email: String): Long = withContext(Dispatchers.IO) {
        val normalizedEmail = email.trim().lowercase(Locale.ROOT)
        dao.getUserIdByEmail(normalizedEmail) ?: -1L
    }

    suspend fun getUserProfile(userId: Long): UserProfileEntity? = withContext(Dispatchers.IO) {
        dao.getUserProfile(userId)
    }

    suspend fun getActiveChapters(): List<String> = withContext(Dispatchers.IO) {
        ensureSeeded()
        dao.getActiveChapters()
    }

    suspend fun saveAIQuestion(userId: Long, question: String, imageUri: String?, aiAnswer: String) = withContext(Dispatchers.IO) {
        val ai = AIQuestionEntity(
            userId = userId,
            questionText = question,
            imageUri = imageUri,
            aiAnswer = aiAnswer,
            createdAt = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US).format(Date())
        )
        dao.insertAIQuestion(ai)
    }

    suspend fun saveQuizAttempt(userId: Long, chapterId: Long, total: Int, correct: Int) = withContext(Dispatchers.IO) {
        val attempt = QuizAttemptEntity(
            userId = userId,
            chapterId = chapterId,
            totalQuestions = total,
            correctAnswers = correct,
            score = if (total > 0) (correct * 100) / total else 0,
            status = "COMPLETED",
            completedAt = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US).format(Date())
        )
        dao.insertQuizAttempt(attempt)
    }

    suspend fun getAIHistory(userId: Long): List<QAItem> = withContext(Dispatchers.IO) {
        val list = dao.getAIHistory(userId)
        list.map { ai ->
            QAItem(
                id = ai.aiQuestionId.toInt(),
                question = ai.questionText,
                imageUri = ai.imageUri,
                answer = ai.aiAnswer,
                timestamp = 0
            )
        }
    }

    suspend fun deleteAIQuestion(aiQuestionId: Long): Boolean = withContext(Dispatchers.IO) {
        dao.deleteAIQuestion(aiQuestionId) > 0
    }

    suspend fun clearAIHistory(userId: Long): Boolean = withContext(Dispatchers.IO) {
        dao.clearAIHistory(userId) > 0
    }

    suspend fun updateXP(userId: Long, xpGain: Int) = withContext(Dispatchers.IO) {
        dao.updateXP(userId, xpGain)
    }

    suspend fun updateMana(userId: Long, manaGain: Int) = withContext(Dispatchers.IO) {
        dao.updateMana(userId, manaGain)
    }

    suspend fun getOrCreateDailyTasks(userId: Long, taskDate: String): List<DailyTaskEntity> = withContext(Dispatchers.IO) {
        val existing = dao.getDailyTasks(userId, taskDate)
        if (existing.isNotEmpty()) return@withContext existing

        insertDailyTask(userId, taskDate, "LESSON_COMPLETE", 1, 50)
        insertDailyTask(userId, taskDate, "PERFECT_SCORE", 1, 100)
        insertDailyTask(userId, taskDate, "ANSWER_QUESTIONS", 10, 50)
        dao.getDailyTasks(userId, taskDate)
    }

    private suspend fun insertDailyTask(userId: Long, taskDate: String, taskType: String, targetValue: Int, rewardXp: Int) {
        val task = DailyTaskEntity(
            userId = userId,
            taskDate = taskDate,
            taskType = taskType,
            title = taskType,
            targetValue = targetValue,
            currentValue = 0,
            rewardXp = rewardXp,
            status = "IN_PROGRESS"
        )
        dao.insertDailyTask(task)
    }

    suspend fun updateDailyTaskProgress(userId: Long, taskDate: String, taskType: String, increment: Int) = withContext(Dispatchers.IO) {
        val tasks = dao.getDailyTasks(userId, taskDate)
        val target = tasks.firstOrNull { it.taskType == taskType }
        if (target != null) {
            dao.updateTaskProgress(target.taskId, increment)
        }
    }

    suspend fun getDailyTask(taskId: Long): DailyTaskEntity? = withContext(Dispatchers.IO) {
        dao.getDailyTask(taskId)
    }

    suspend fun claimDailyTask(taskId: Long): Boolean = withContext(Dispatchers.IO) {
        dao.claimDailyTask(taskId) > 0
    }

    private fun chestKey(userId: Long, taskDate: String) = "chest_${userId}_$taskDate"

    fun isDailyChestClaimed(userId: Long, taskDate: String): Boolean {
        return context.getSharedPreferences("StudyMentorDailyState", Context.MODE_PRIVATE)
            .getBoolean(chestKey(userId, taskDate), false)
    }

    @Synchronized
    fun claimDailyChest(userId: Long, taskDate: String): Boolean {
        val prefs = context.getSharedPreferences("StudyMentorDailyState", Context.MODE_PRIVATE)
        val key = chestKey(userId, taskDate)
        if (prefs.getBoolean(key, false)) return false
        return prefs.edit().putBoolean(key, true).commit()
    }

    private fun mapNodeKey(userId: Long, chapterTitle: String) = "node_${userId}_$chapterTitle"

    fun getCurrentLessonNode(userId: Long, chapterTitle: String): Int {
        return context.getSharedPreferences("StudyMentorMapProgress", Context.MODE_PRIVATE)
            .getInt(mapNodeKey(userId, chapterTitle), 1)
    }

    fun saveCurrentLessonNode(userId: Long, chapterTitle: String, node: Int) {
        val safeNode = node.coerceIn(1, 6)
        context.getSharedPreferences("StudyMentorMapProgress", Context.MODE_PRIVATE)
            .edit()
            .putInt(mapNodeKey(userId, chapterTitle), safeNode)
            .apply()
    }

    suspend fun getLeaderboard(): List<UserProfileEntity> = withContext(Dispatchers.IO) {
        dao.getLeaderboard()
    }

    suspend fun updateStreak(userId: Long, streak: Int, lastActiveDate: String): Boolean = withContext(Dispatchers.IO) {
        dao.updateStreak(userId, streak, lastActiveDate) > 0
    }

    suspend fun updateProfileInfo(userId: Long, displayName: String, avatarUri: String?, motto: String?): Boolean = withContext(Dispatchers.IO) {
        dao.updateProfileInfo(userId, displayName, avatarUri, motto) > 0
    }

    suspend fun insertVocabulary(userId: Long, word: String, phonetic: String, meaning: String, example: String): Long = withContext(Dispatchers.IO) {
        val vocab = VocabularyEntity(
            userId = userId,
            word = word,
            phonetic = phonetic,
            meaning = meaning,
            example = example,
            createdAt = SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault()).format(Date())
        )
        dao.insertVocabulary(vocab)
    }

    suspend fun getVocabularyList(userId: Long): List<VocabularyEntity> = withContext(Dispatchers.IO) {
        dao.getVocabularyList(userId)
    }

    suspend fun toggleVocabularyMastered(vocabId: Long, isMastered: Boolean): Boolean = withContext(Dispatchers.IO) {
        dao.updateVocabularyMastered(vocabId, isMastered) > 0
    }

    suspend fun deleteVocabulary(vocabId: Long): Boolean = withContext(Dispatchers.IO) {
        dao.deleteVocabulary(vocabId) > 0
    }

    suspend fun isWordBookmarked(userId: Long, word: String): Boolean = withContext(Dispatchers.IO) {
        dao.isWordBookmarked(userId, word) > 0
    }
}

